File: bitshuriken-prod-match/messaging/topics.py
"""Kafka 토픽 이름과 op 코드 상수."""

# Op codes
OP_NEW_ORDER = "NO"
OP_CANCEL_ORDER = "CO"
OP_TRADE = "TR"
OP_ORDER_UPDATE = "OU"
OP_DEPTH_DIFF = "DPD"

# Control topic ops (런타임 ticker 상장)
OP_ADD_TICKER = "ADD"
CONTROL_PARTITION = 0  # control 토픽은 1 partition (compacted, key=symbol)

# Market slugs (Prisma MarketType의 lowercase)
MARKET_SPOT = "spot"
MARKET_FUTURES = "futures"


def inbound_topic(market: str) -> str:
    return f"match.{market}.in"


def outbound_topic(market: str) -> str:
    return f"match.{market}.out"


def book_topic(market: str) -> str:
    """Orderbook diff stream용 별도 topic."""
    return f"match.{market}.book"


def state_topic(market: str) -> str:
    """Lane 스냅샷용 log-compacted topic. key=symbol."""
    return f"match.{market}.state"


def control_topic(market: str) -> str:
    """Ticker 라이프사이클 컨트롤용 log-compacted topic. key=symbol, 1 partition."""
    return f"match.{market}.control"

File: bitshuriken-prod-be/libs/infra/src/messaging/topics.ts
import { MarketType } from '@prisma/client';

export const Op = {
  NEW_ORDER: 'NO',
  CANCEL_ORDER: 'CO',
  TRADE: 'TR',
  ORDER_UPDATE: 'OU',
  DEPTH_DIFF: 'DPD',
} as const;

/** Control topic ops (런타임 ticker 상장). */
export const ControlOp = {
  ADD_TICKER: 'ADD',
} as const;

/** Control 토픽은 1 partition(compacted, key=symbol). */
export const CONTROL_PARTITION = 0;

export type OpCode = (typeof Op)[keyof typeof Op];

const MARKET_SLUG: Record<MarketType, string> = {
  SPOT: 'spot',
  FUTURES: 'futures',
};

export function inboundTopic(market: MarketType): string {
  return `match.${MARKET_SLUG[market]}.in`;
}

export function outboundTopic(market: MarketType): string {
  return `match.${MARKET_SLUG[market]}.out`;
}

/** Orderbook diff stream용 별도 topic. */
export function bookTopic(market: MarketType): string {
  return `match.${MARKET_SLUG[market]}.book`;
}

/** Ticker 라이프사이클 컨트롤용 log-compacted topic. */
export function controlTopic(market: MarketType): string {
  return `match.${MARKET_SLUG[market]}.control`;
}

export const ALL_OUTBOUND_TOPICS: string[] = (Object.keys(MARKET_SLUG) as MarketType[]).map(
  outboundTopic,
);
